File: src/main/java/ini/cx3d/physics/interfaces/IntracellularSubstance.java
package ini.cx3d.physics.interfaces;

/**
 * Created by lukas on 25.02.16.
 */
public interface IntracellularSubstance extends Substance {
	/**
	 * Distribute IntracellularSubstance concentration at division and update quantity.
	 * @param newIS
	 */
	void distributeConcentrationOnDivision(IntracellularSubstance newIS);

	/**
	 * Degradation of the <code>IntracellularSubstance</code>.
	 * @param newIS
	 */
	void degrade();

	/** Returns the degree of asymmetric distribution during cell division.
	 * 0 represents complete symmetrical division, 1 represents complete asymmetric division. */
	double getAsymmetryConstant();

	/** Sets the degree of asymmetric distribution during cell division.
	 * 0 represents complete symmetrical division, 1 represents complete asymmetric division.
	 * The sign + or - is used to distinguish between one daughter (mother cell) and the other
	 * (new cell). */
	void setAsymmetryConstant(double asymmetryConstant);

	/** If true, the Substance can be detected from outside of the PhysicalObject
	 * (equivalent to an membrane bound substance).*/
	boolean isVisibleFromOutside();

	/** If true, the Substance can be detected from outside of the PhysicalObject
	 * (equivalent to an membrane bound substance).*/
	void setVisibleFromOutside(boolean visibleFromOutside);

	/** If true, the volume is taken into account for computing the concentration,
	 * otherwise a virtual volume corresponding to the length of the physical object
	 * (with virtual radius 1) is used.*/
	boolean isVolumeDependant();

	/** If true, the volume is taken into account for computing the concentration,
	 * otherwise a virtual volume corresponding to the length of the physical object
	 * (with virtual radius 1) is used.*/
	void setVolumeDependant(boolean volumeDependant);
}

File: src/physics/intracellular_substance.cc
#include "physics/intracellular_substance.h"

#include "sim_state_serialization_util.h"

namespace cx3d {
namespace physics {

IntracellularSubstance::IntracellularSubstance()
    : Substance(),
      asymmetry_constant_ { 0.0 },
      visible_from_outside_ { false },
      volume_dependant_ { false } {

}

IntracellularSubstance::IntracellularSubstance(const std::string& id, double diffusion_constant,
                                               double degradation_constant)
    : Substance(id, diffusion_constant, degradation_constant),
      asymmetry_constant_ { 0.0 },
      visible_from_outside_ { false },
      volume_dependant_ { false } {

}

IntracellularSubstance::IntracellularSubstance(const IntracellularSubstance& other)
    : Substance(other),
      asymmetry_constant_ { other.asymmetry_constant_ },
      visible_from_outside_ { other.visible_from_outside_ },
      volume_dependant_ { other.volume_dependant_ } {
}

StringBuilder& IntracellularSubstance::simStateToJson(StringBuilder& sb) const {
  Substance::simStateToJson(sb);
  SimStateSerializationUtil::removeLastChar(sb);
  sb.append(",");

  SimStateSerializationUtil::keyValue(sb, "visibleFromOutside", visible_from_outside_);
  SimStateSerializationUtil::keyValue(sb, "volumeDependant", volume_dependant_);
  SimStateSerializationUtil::keyValue(sb, "asymmetryConstant", asymmetry_constant_);

  SimStateSerializationUtil::removeLastChar(sb);
  sb.append("}");
  return sb;
}

void IntracellularSubstance::distributeConcentrationOnDivision(IntracellularSubstance* new_is) {
  double p = asymmetry_constant_ * (1 - concentration_) * concentration_;
  new_is->setConcentration(concentration_ * (1 + p));
  concentration_ = concentration_ * (1 - p);
}

void IntracellularSubstance::degrade() {
  concentration_ = concentration_ * (1 - degradation_constant_ * Param::kSimulationTimeStep);
}

double IntracellularSubstance::getAsymmetryConstant() const {
  return asymmetry_constant_;
}

void IntracellularSubstance::setAsymmetryConstant(double asymmetry_constant) {
  asymmetry_constant_ = asymmetry_constant;
}

bool IntracellularSubstance::isVisibleFromOutside() const {
  return visible_from_outside_;
}

void IntracellularSubstance::setVisibleFromOutside(bool visible_from_outside) {
  visible_from_outside_ = visible_from_outside;
}

bool IntracellularSubstance::isVolumeDependant() const {
  return volume_dependant_;
}

void IntracellularSubstance::setVolumeDependant(bool volume_dependant) {
  volume_dependant_ = volume_dependant;
}

Substance::UPtr IntracellularSubstance::getCopy() const {
  return Substance::UPtr(new IntracellularSubstance(*this));
}

}  // namespace physics
}  // namespace cx3d
